lib/src/migration_visitor.dart

```diff
@@ -0,0 +1,101 @@
+// Copyright 2019 Google LLC
+//
+// Use of this source code is governed by an MIT-style
+// license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import 'dart:io';
+
+// The sass package's API is not necessarily stable. It is being imported with
+// the Sass team's explicit knowledge and approval. See
+// https://github.com/sass/dart-sass/issues/236.
+import 'package:sass/src/ast/sass.dart';
+import 'package:sass/src/syntax.dart';
+import 'package:sass/src/visitor/recursive_ast.dart';
+
+import 'package:path/path.dart' as p;
+
+import 'migrator.dart';
+import 'patch.dart';
+import 'utils.dart';
+
+/// A visitor that migrates a stylesheet.
+///
+/// When [run] is called, this visitor traverses a stylesheet's AST, allowing
+/// subclasses to override one or more methods and add to [patches]. Once the
+/// stylesheet has been visited, the migrated contents (based on [patches]) will
+/// be stored in [migrator]'s [migrated] map.
+///
+/// If [migrateDependencies] is enabled, this visitor will construct and run a
+/// new instance of itself (using [newInstance]) each time it encounters an
+/// @import or @use rule.
+abstract class MigrationVisitor extends RecursiveAstVisitor {
+  /// The migrator running on this stylesheet.
+  Migrator get migrator;
+
+  /// The canonical path of the stylesheet being migrated.
+  String get path;
+
+  /// The stylesheet being migrated.
+  Stylesheet stylesheet;
+
+  /// The syntax this stylesheet uses.
+  Syntax syntax;
+
+  /// The patches to be applied to the stylesheet being migrated.
+  final List<Patch> patches = [];
+
+  /// Returns a new instance of this MigrationVisitor with the same migrator
+  /// and [newPath].
+  MigrationVisitor newInstance(String newPath);
+
+  /// Runs the migrator and stores the migrated contents in `migrator.migrated`.
+  void run() {
+    var contents = File(path).readAsStringSync();
+    syntax = Syntax.forPath(path);
+    stylesheet = Stylesheet.parse(contents, syntax, url: path);
+    visitStylesheet(stylesheet);
+    var results = getMigratedContents();
+    if (results != null) {
+      migrator.migrated[path] = results;
```

Comment:
Changed this to have `migrated` as a field of `MigrationVisitor`, with any dependency visitors created with `newInstance` getting a reference to the same map so that they can all add to it and the entrypoint's visitor can pass the complete map back to `Migrator` when `run` returns.

Do you think this is fine, or would it be better to have each visitor maintain its own map and then copy the results from each of its dependencies?

Reply:
That's a good question. I'd say make the `migrated` field private to `MigrationVisitor` and give it total responsibility for handling the output, but you need to pass it through `newInstance()` somehow...

What if we restructure this so that you don't need to construct multiple visitors at all? Rather than passing a URL to the constructor, pass it to `run()`. That also means that when we get around to de-duplicating repeated imports/uses (which will probably be necessary for large-scale performance), there's a natural place for that as well.

This does mean that per-file state can't just be stored as final fields on the subclass. But you can override `visitStylesheet()` to set it temporarily:

```dart
@override
void visitStylesheet(StylesheetNode node) {
  var oldNamespaces = _namespaces;
  var oldAdditionalUseRules = _additionalUseRules;
  var oldConfigurableVariables = _configurableVariables;
  _namespaces = {};
  _additionalUseRules = Set();
  _configurableVariables = normalizedSet();
  super.visitStylesheet(node);
  _namespaces = oldNamespaces;
  _additionalUseRules = oldAdditionalUseRules;
  _configurableVariables = oldConfigurableVariables;
}
```

This is basically how Dart Sass's `EvaluateVisitor` works. WDYT?